From f3e9b66f0797ceb50f60038f353d757d99d61eec Mon Sep 17 00:00:00 2001
From: Vasilis Liaskovitis <vliaskovitis@suse.com>
Date: Sun, 18 Oct 2020 00:35:07 +0200
Subject: [PATCH 01/10] [PATCH 01/11] modesetting: make drmmode_shadow_allocate
 more generic

To do so, introduce a drmmode_scanout_rec structure that contains:
 - the DRM BO
 - the DRM FB id

We then take the generic part of drmmode_shadow_allocate that creates a
scanable buffer and move it to drmmode_scanout_allocate().

Signed-off-by: Martin Peres <martin.peres@linux.intel.com>
---
 hw/xfree86/drivers/modesetting/dri2.c         |  2 +-
 .../drivers/modesetting/drmmode_display.c     | 60 ++++++++++++-------
 .../drivers/modesetting/drmmode_display.h     |  8 ++-
 hw/xfree86/drivers/modesetting/present.c      |  2 +-
 4 files changed, 46 insertions(+), 26 deletions(-)

diff --git a/hw/xfree86/drivers/modesetting/dri2.c b/hw/xfree86/drivers/modesetting/dri2.c
index c0799896c..06534e173 100644
--- a/hw/xfree86/drivers/modesetting/dri2.c
+++ b/hw/xfree86/drivers/modesetting/dri2.c
@@ -545,7 +545,7 @@ can_exchange(ScrnInfoPtr scrn, DrawablePtr draw,
 
         /* Don't do pageflipping if CRTCs are rotated. */
 #ifdef GLAMOR_HAS_GBM
-        if (drmmode_crtc->rotate_bo.gbm)
+        if (drmmode_crtc->shadow_rotated.bo.gbm)
             return FALSE;
 #endif
 
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index 9dd8c5573..cba45775d 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -617,8 +617,8 @@ drmmode_crtc_get_fb_id(xf86CrtcPtr crtc, uint32_t *fb_id, int *x, int *y)
             *x = drmmode_crtc->prime_pixmap_x;
         *y = 0;
     }
-    else if (drmmode_crtc->rotate_fb_id) {
-        *fb_id = drmmode_crtc->rotate_fb_id;
+    else if (drmmode_crtc->shadow_rotated.fb_id) {
+        *fb_id = drmmode_crtc->shadow_rotated.fb_id;
         *x = *y = 0;
     }
     else {
@@ -1810,34 +1810,50 @@ drmmode_clear_pixmap(PixmapPtr pixmap)
     }
 }
 
-static void *
-drmmode_shadow_allocate(xf86CrtcPtr crtc, int width, int height)
+static Bool
+drmmode_scanout_allocate(xf86CrtcPtr crtc, int width, int height,
+                         drmmode_shadow_scanout_ptr scanout)
 {
     drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
     drmmode_ptr drmmode = drmmode_crtc->drmmode;
     int ret;
 
-    if (!drmmode_create_bo(drmmode, &drmmode_crtc->rotate_bo,
+    if (!drmmode_create_bo(drmmode, &scanout->bo,
                            width, height, drmmode->kbpp)) {
         xf86DrvMsg(crtc->scrn->scrnIndex, X_ERROR,
-               "Couldn't allocate shadow memory for rotated CRTC\n");
-        return NULL;
+                   "Couldn't allocate scanout memory\n");
+        return FALSE;
     }
 
-    ret = drmmode_bo_import(drmmode, &drmmode_crtc->rotate_bo,
-                            &drmmode_crtc->rotate_fb_id);
+    ret = drmmode_bo_import(drmmode, &scanout->bo,
+                            &scanout->fb_id);
 
     if (ret) {
-        ErrorF("failed to add rotate fb\n");
-        drmmode_bo_destroy(drmmode, &drmmode_crtc->rotate_bo);
-        return NULL;
+        ErrorF("failed to add scanout fb\n");
+        drmmode_bo_destroy(drmmode, &scanout->bo);
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+static void *
+drmmode_shadow_allocate(xf86CrtcPtr crtc, int width, int height)
+{
+    drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
+    drmmode_ptr drmmode = drmmode_crtc->drmmode;
+
+    if (!drmmode_scanout_allocate(crtc, width, height,
+                                  &drmmode_crtc->shadow_rotated)) {
+        xf86DrvMsg(crtc->scrn->scrnIndex, X_ERROR,
+                   "Couldn't allocate shadow memory for rotated CRTC\n");
     }
 
 #ifdef GLAMOR_HAS_GBM
     if (drmmode->gbm)
-        return drmmode_crtc->rotate_bo.gbm;
+        return drmmode_crtc->shadow_rotated.bo.gbm;
 #endif
-    return drmmode_crtc->rotate_bo.dumb;
+    return drmmode_crtc->shadow_rotated.bo.dumb;
 }
 
 static PixmapPtr
@@ -1881,14 +1897,14 @@ drmmode_shadow_create(xf86CrtcPtr crtc, void *data, int width, int height)
         }
     }
 
-    if (!drmmode_bo_has_bo(&drmmode_crtc->rotate_bo)) {
+    if (!drmmode_bo_has_bo(&drmmode_crtc->shadow_rotated.bo)) {
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Couldn't allocate shadow pixmap for rotated CRTC\n");
         return NULL;
     }
 
-    pPixData = drmmode_bo_map(drmmode, &drmmode_crtc->rotate_bo);
-    rotate_pitch = drmmode_bo_get_pitch(&drmmode_crtc->rotate_bo),
+    pPixData = drmmode_bo_map(drmmode, &drmmode_crtc->shadow_rotated.bo);
+    rotate_pitch = drmmode_bo_get_pitch(&drmmode_crtc->shadow_rotated.bo),
 
     rotate_pixmap = drmmode_create_pixmap_header(scrn->pScreen,
                                                  width, height,
@@ -1903,7 +1919,7 @@ drmmode_shadow_create(xf86CrtcPtr crtc, void *data, int width, int height)
         return NULL;
     }
 
-    drmmode_set_pixmap_bo(drmmode, rotate_pixmap, &drmmode_crtc->rotate_bo);
+    drmmode_set_pixmap_bo(drmmode, rotate_pixmap, &drmmode_crtc->shadow_rotated.bo);
 
     return rotate_pixmap;
 }
@@ -1919,11 +1935,11 @@ drmmode_shadow_destroy(xf86CrtcPtr crtc, PixmapPtr rotate_pixmap, void *data)
     }
 
     if (data) {
-        drmModeRmFB(drmmode->fd, drmmode_crtc->rotate_fb_id);
-        drmmode_crtc->rotate_fb_id = 0;
+        drmModeRmFB(drmmode->fd, drmmode_crtc->shadow_rotated.fb_id);
+        drmmode_crtc->shadow_rotated.fb_id = 0;
 
-        drmmode_bo_destroy(drmmode, &drmmode_crtc->rotate_bo);
-        memset(&drmmode_crtc->rotate_bo, 0, sizeof drmmode_crtc->rotate_bo);
+        drmmode_bo_destroy(drmmode, &drmmode_crtc->shadow_rotated.bo);
+        memset(&drmmode_crtc->shadow_rotated.bo, 0, sizeof drmmode_crtc->shadow_rotated.bo);
     }
 }
 
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index 4142607fb..6f3643f3e 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -153,6 +153,11 @@ typedef struct {
     uint64_t *modifiers;
 } drmmode_format_rec, *drmmode_format_ptr;
 
+typedef struct {
+    drmmode_bo bo;
+    unsigned fb_id;
+} drmmode_shadow_scanout_rec, *drmmode_shadow_scanout_ptr;
+
 typedef struct {
     drmmode_ptr drmmode;
     drmModeCrtcPtr mode_crtc;
@@ -169,8 +174,7 @@ typedef struct {
     uint32_t num_formats;
     drmmode_format_rec *formats;
 
-    drmmode_bo rotate_bo;
-    unsigned rotate_fb_id;
+    drmmode_shadow_scanout_rec shadow_rotated;
 
     PixmapPtr prime_pixmap;
     PixmapPtr prime_pixmap_back;
diff --git a/hw/xfree86/drivers/modesetting/present.c b/hw/xfree86/drivers/modesetting/present.c
index 009a0790c..80ca19c67 100644
--- a/hw/xfree86/drivers/modesetting/present.c
+++ b/hw/xfree86/drivers/modesetting/present.c
@@ -241,7 +241,7 @@ ms_present_check_unflip(RRCrtcPtr crtc,
         drmmode_crtc_private_ptr drmmode_crtc = config->crtc[i]->driver_private;
 
         /* Don't do pageflipping if CRTCs are rotated. */
-        if (drmmode_crtc->rotate_bo.gbm)
+        if (drmmode_crtc->shadow_rotated.bo.gbm)
             return FALSE;
 
         if (ms_crtc_on(config->crtc[i]))
-- 
2.25.1

